README.md
# kintone-follow-people

NPM package to follow people in kintone

[![npm version](https://img.shields.io/npm/v/kintone-follow-people.svg)](https://www.npmjs.com/package/kintone-follow-people)
[![npm downloads](https://img.shields.io/npm/dm/kintone-follow-people.svg)](https://www.npmjs.com/package/kintone-follow-people)
![Node.js Version Support](https://img.shields.io/badge/Node.js%20support-v4–v7-brightgreen.svg)
[![Build Status](https://travis-ci.org/miyajan/kintone-follow-people.svg?branch=master)](https://travis-ci.org/miyajan/kintone-follow-people)
[![dependencies Status](https://david-dm.org/miyajan/kintone-follow-people/status.svg)](https://david-dm.org/miyajan/kintone-follow-people)
[![Coverage Status](https://coveralls.io/repos/github/miyajan/kintone-follow-people/badge.svg?branch=master)](https://coveralls.io/github/miyajan/kintone-follow-people?branch=master)
![License](https://img.shields.io/npm/l/kintone-follow-people.svg)

## Description

This tool help to follow people in kintone.

## Install

```
$ npm install kintone-follow-people -g
```

## Usage

Follow all people in your kintone!

```
$ kintone-follow-people -f example.cybozu.com -u your-login-name -p your-password
```

## Options

### ```-f``` or ```--fqdn```

Specify FQDN of your kintone's domain.

### ```-u``` or ```--username```

Specify your login name.

### ```-p``` or ```--password```

Specify your password.

### ```-e``` or ```--excludes```

Specify users not to follow by comma-separated list of login names.

### ```-i``` or ```--includes```

If this options is specified, follow only users included in this option. Specify by comma-separated list of login names.

### ```-c``` or ```--config```

Specify a path to configuration file. You can load above options from the json file like:

```json
{
    "fqdn": "example.cybozu.com",
    "username": "your-login-name",
    "password": "your-password",
    "excludes": [
        "non-follow-user1",
        "non-follow-user2"
    ],
    "includes": [
        "follow-user1",
        "follow-user2"
    ]
}
```

## Contribution

1. Fork
2. Create a feature branch
3. Commit your changes
4. Rebase your local changes against the master branch
5. Run `npm test`
6. Create new Pull Request

## License

MIT

## Author

[miyajan](https://github.com/miyajan): Jumpei Miyata miyajan777@gmail.com


CHANGELOG.md
<a name="0.3.0"></a>
# [0.3.0](https://github.com/miyajan/kintone-follow-people/compare/0.2.0...0.3.0) (2018-03-28)


### Features

* support env proxy (http_proxy, https_proxy and so on) ([704f0e6](https://github.com/miyajan/kintone-follow-people/commit/704f0e6))



<a name="0.2.0"></a>
# [0.2.0](https://github.com/miyajan/kintone-follow-people/compare/0.1.0...0.2.0) (2017-01-19)


#### Features

* **options:**
  * add --includes option to follow only specified people ([029dde8d](https://github.com/miyajan/kintone-follow-people/commit/029dde8dcc233f6f8e01e99cce9eb43c46d463e8))


## 0.1.0 (2017-01-17)

* First release!


test/unit/index.js
'use strict';

const proxyquire = require('proxyquire');
const assert = require('assert');

describe('followPeople', function() {
    const fqdn = 'test.cybozu.com';
    const username = 'test-code1';
    const password = 'test-password';
    let sut, usersData, subscribeData;

    beforeEach(function() {
        usersData = [];
        subscribeData = [];
        sut = proxyquire('../../index', {
            'fetch-with-proxy': {
                default: (url, opts) => {
                    if (url === `https://${fqdn}/k/api/group/users.json`) {
                        usersData.push(opts.body);
                        return Promise.resolve({
                            status: 200,
                            json: () => Promise.resolve({
                                result: {
                                    entities: [
                                        {
                                            code: 'test-code1',
                                            name: 'test-name1',
                                            id: 'test-id1'
                                        },
                                        {
                                            code: 'test-code2',
                                            name: 'test-name2',
                                            id: 'test-id2'
                                        },
                                        {
                                            code: 'test-code3',
                                            name: 'test-name3',
                                            id: 'test-id3'
                                        }
                                    ]
                                }
                            })
                        });
                    } else if (url === `https://${fqdn}/k/api/people/user/subscribe.json`) {
                        subscribeData.push(opts.body);
                        return Promise.resolve({
                            status: 200,
                            json: () => Promise.resolve()
                        })
                    } else {
                        assert(false);
                    }
                }
            }
        });
    });

    it('follow people', function() {
        return sut(fqdn, username, password).then(users => {
            assert.deepEqual(usersData, [JSON.stringify({
                'id': '7532782697181632513',
                'size': 1000,
                'offset': 0
            })]);
            assert.deepEqual(subscribeData, [
                JSON.stringify({
                    'userId': 'test-id2',
                    'subscribe': true
                }),
                JSON.stringify({
                    'userId': 'test-id3',
                    'subscribe': true
                })
            ]);
            assert.deepEqual(users, [
                {
                    code: 'test-code2',
                    name: 'test-name2',
                    id: 'test-id2'
                },
                {
                    code: 'test-code3',
                    name: 'test-name3',
                    id: 'test-id3'
                }
            ]);
        });
    });

    it('not follow people included in --excludes option', function() {
        return sut(fqdn, username, password, ['test-code3']).then(users => {
            assert.deepEqual(subscribeData, [
                JSON.stringify({
                    'userId': 'test-id2',
                    'subscribe': true
                })
            ]);
            assert.deepEqual(users, [
                {
                    code: 'test-code2',
                    name: 'test-name2',
                    id: 'test-id2'
                }
            ]);
        });
    });

    it('only follow people included in --includes option if specified', function() {
        return sut(fqdn, username, password, [], ['test-code3']).then(users => {
            assert.deepEqual(subscribeData, [
                JSON.stringify({
                    'userId': 'test-id3',
                    'subscribe': true
                })
            ]);
            assert.deepEqual(users, [
                {
                    code: 'test-code3',
                    name: 'test-name3',
                    id: 'test-id3'
                }
            ]);
        });
    });
});


lib/cli.js
'use strict';

const commander = require('commander');
const followPeople = require('../index');
const pkg = require('../package.json');

/**
 * split value by comma
 * @param {string} value
 * @return {!Array.<string>}
 */
function separate(value) {
    return value.split(',');
}

/**
 * Main
 * @param {!Array} argv Arguments
 * @param {!Console} console Console
 * @param {!Function} exit Function to exit
 */
function main(argv, console, exit) {
    const program = new commander.Command();

    program
        .version(pkg.version, '-v, --version')
        .option('-f, --fqdn <fqdn>', 'FQDN for kintone')
        .option('-u, --username <username>', 'Username for kintone')
        .option('-p, --password <password>', 'Password for kintone')
        .option('-e, --excludes <codes>', 'Users not to subscribe. Comma separated list of login name.', separate)
        .option('-i, --includes <codes>', 'Follow only specified users. Comma separeted list of login name.', separate)
        .option('-c, --config <path>', 'Path to config json file');

    program.parse(argv);

    let config = {};
    const configPath = program.config;
    if (configPath) {
        try {
            config = require(configPath);
        } catch (e) {
            console.error(`Error! Can't read config json: ${configPath}`);
            throw e;
        }
    }

    const fqdn = program.fqdn || config.fqdn;
    const username = program.username || config.username;
    const password = program.password || config.password;
    const excludes = program.excludes || config.excludes;
    const includes = program.includes || config.includes;

    if (!fqdn || !username || !password) {
        program.outputHelp();
        exit(1);
    }

    followPeople(fqdn, username, password, excludes, includes).then(users => {
        console.log(`Done! Now following ${users.length} users.`);
    }).catch(err => {
        console.error(err.message);
        exit(1);
    });
}

module.exports = main;


index.js
'use strict';

const fetch = require('fetch-with-proxy').default;
const Base64 = require('js-base64').Base64;

const USER_PATH = '/k/api/group/users.json';
const SUBSCRIBE_PATH = '/k/api/people/user/subscribe.json';
const EVERYONE_GROUP_ID = '7532782697181632513';
const MAX_GET_USERS_SIZE = 1000;

class KintoneFollowPeople {
    /**
     * Constructor
     * @param {string} fqdn FQDN
     * @param {string} username Username
     * @param {string} password Password
     * @param {Array.<string>=} opt_excludes Codes of users not to follow
     * @param {Array.<string>=} opt_includes Only follow users whose code included in this parameter
     */
    constructor(fqdn, username, password, opt_excludes, opt_includes) {
        this._fqdn = fqdn;
        this._username = username;
        this._password = password;
        this._excludes = opt_excludes || [];
        this._includes = opt_includes || [];
    }

    /**
     * Execute to follow peoples
     * @return {!Thenable.<!Array.<!Object>>} Thenable. Users are passed to then.
     */
    execute() {
        return this._getAllUsers().then(users => {
            users = users.filter(user => this._doSubscribe(user));
            let promise = Promise.resolve();
            users.forEach(user => {
                promise = promise.then(() => {
                    return this._subscribeUser(user);
                });
            });
            return promise.then(() => users);
        });
    }

    /**
     * @param {string} path Path
     * @param {!Object} data Post data
     * @return {!Thenable}
     * @private
     */
    _postJson(path, data) {
        return fetch('https://' + this._fqdn + path, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'X-Cybozu-Authorization': Base64.encode(this._username + ':' + this._password)
            },
            body: JSON.stringify(data)
        }).then(response => {
            if (response.status >= 200 && response.status < 300) {
                return response.json();
            } else {
                return Promise.reject(new Error(response.statusText));
            }
        });
    }

    /**
     * @param {!Array.<!Object>} users Users
     * @param {number} offset Offset
     * @return {!Thenable}
     * @private
     */
    _getUsersRecursively(users, offset) {
        return this._postJson(USER_PATH, {
            'id': EVERYONE_GROUP_ID,
            'size': MAX_GET_USERS_SIZE,
            'offset': offset,
        }).then(resp => {
            users = users.concat(resp.result.entities);
            if (users.length < offset + MAX_GET_USERS_SIZE) {
                return users;
            }
            return this._getUsersRecursively(users, offset + MAX_GET_USERS_SIZE);
        });
    }

    /**
     * @return {!Thenable}
     * @private
     */
    _getAllUsers() {
        return this._getUsersRecursively([], 0);
    }

    /**
     * @param {!Object} user User
     * @return {!Thenable}
     * @private
     */
    _subscribeUser(user) {
        console.log('Following... ' + user.name);
        return this._postJson(SUBSCRIBE_PATH, {
            'userId': user.id,
            'subscribe': true,
        });
    }

    /**
     * @param {!Object} user User
     * @return {boolean}
     * @private
     */
    _doSubscribe(user) {
        // not subscribe myself
        if (user.code === this._username) {
            return false;
        }
        // not subscribe users included in --excludes option
        if (this._excludes.indexOf(user.code) > -1) {
            return false;
        }
        // not subscribe users not included in --includes option if specified
        if (this._includes.length > 0 && this._includes.indexOf(user.code) === -1) {
            return false;
        }
        // subscribe
        return true;
    }
}

/**
 * Follow people
 * @param {string} fqdn FQDN
 * @param {string} username Username
 * @param {string} password Password
 * @param {Array.<string>=} opt_excludes Codes of users not to follow
 * @param {Array.<string>=} opt_includes Only follow users whose code included in this parameter
 * @return {!Thenable.<!Array.<!Object>>} Thenable. Users are passed to then.
 */
const followPeople = (fqdn, username, password, opt_excludes, opt_includes) => {
    return new KintoneFollowPeople(fqdn, username, password, opt_excludes, opt_includes).execute();
};

module.exports = followPeople;
